cmd2/cmd2.py

```diff
@@ -2959,160 +2959,159 @@ def do_py(self, args: argparse.Namespace) -> bool:
-
-                # Set up sys module for the Python console
-                self._reset_py_display()
-                keepstate = Statekeeper(sys, ('stdin', 'stdout'))
-                sys.stdout = self.stdout
-                sys.stdin = self.stdin
-
-                cprt = 'Type "help", "copyright", "credits" or "license" for more information.'
-                instructions = ('End with `Ctrl-D` (Unix) / `Ctrl-Z` (Windows), `quit()`, `exit()`.\n'
-                                'Non-Python commands can be issued with: {}("your command")\n'
-                                'Run Python code from external script files with: run("script.py")'
-                                .format(self.pyscript_name))
-
-                try:
-                    interp.interact(banner="Python {} on {}\n{}\n\n{}\n".
-                                    format(sys.version, sys.platform, cprt, instructions))
-                except EmbeddedConsoleExit:
-                    pass
-
-                finally:
-                    keepstate.restore()
-
-                    # Set up readline for cmd2
-                    if rl_type != RlType.NONE:
-                        # Save py's history
-                        self.py_history.clear()
-                        for i in range(1, readline.get_current_history_length() + 1):
-                            self.py_history.append(readline.get_history_item(i))
+                            rl_basic_quote_characters.value = old_basic_quotes
 
-                        readline.clear_history()
-
-                        # Restore cmd2's history
-                        for item in saved_cmd2_history:
-                            readline.add_history(item)
-
-                        if self.use_rawinput and self.completekey:
-                            # Restore cmd2's tab completion settings
-                            readline.set_completer(old_completer)
-                            readline.set_completer_delims(old_delims)
-
-                            if rl_type == RlType.GNU:
-                                rl_basic_quote_characters.value = old_basic_quotes
-
-                                if 'gnureadline' in sys.modules:
-                                    # Restore what the readline module pointed to
-                                    if saved_readline is None:
-                                        del(sys.modules['readline'])
-                                    else:
-                                        sys.modules['readline'] = saved_readline
+                            if 'gnureadline' in sys.modules:
+                                # Restore what the readline module pointed to
+                                if saved_readline is None:
+                                    del(sys.modules['readline'])
+                                else:
+                                    sys.modules['readline'] = saved_readline
 
-        except Exception:
```

Comment:
I think we need something like this to catch a KeyboardInterrupt exception (Ctrl-C) and let that gracefully exit the interactive Python session and drop us back to the console

Reply:
I allow the interactive Python session to handle their own exceptions. I did however add KeyboardInterrupt suppression to py and pyscript for when the interactive console isn't running.